### PR TITLE
feat: add fx options plugin

### PR DESCRIPTION
**Pull request by guseggert**
This adds a plugin interface that lets the plugin modify the fx
options that are passed to fx when the app is initialized. This means
plugins can inject their own implementations of IPFS interfaces. This
enables granular customization of go-ipfs behavior by plugins, such
as:

- Bitswap with custom filters (e.g. for CID blocking)

- Custom interface implementations such as Pinner or DAGService

- Dynamic configuration of libp2p ...

One downside of this is that we're exposing the entire dependency
graph, init hooks, initialization, etc. to users, so this comes with a
caveat that we reserve the right to make breaking changes to the graph
structure and initialization logic (although this historically happens
rarely). If these things are changed, we should mention them in
release notes and changelogs though, since they could impact users of
this plugin interface.

I'm not particularly fond of DI frameworks (and neither are some of
the folks work on/near go-ipfs), but it seems unlikely that somebody
will rewrite the dependency wiring and lifecycle hooks of go-ipfs, and
add dynamic extension points, so this seems like a palatable
compromise.

There are also problems that we should clean up in how model the
go-ipfs app in fx, such as:

- We make extensive use of nested fx.Options, which fx itself
discourages because it "limits the user's ability to customize their
application". It should be easy to flatten these out into a single
[]fx.Option slice.

- We pass around a list of opaque libp2p opts, which makes it hard to
customize after-the-fact...we should consider naming each of these
opts and providing them to fx as proper dependencies, so that they can
be explicitly overridden.

- We call fx.Invoke() in some places with anonymous functions. We
should instead only pass exported functions to fx.Invoke(), so that
they have exported names, which would make it easier to remove/augment
the invocations that happen when the app is initialized.

These aren't blocking issues, they just make it harder and more
brittle to customize go-ipfs with this plugin.

Closes https://github.com/ipfs/go-ipfs/issues/7653

### CODE REVIEW

**Comment by guseggert** [comment]
To elaborate on the plugin interface: the reason it takes the full list of fx options and returns the full list, is to maximize the ability to customize what gets passed to fx. Currently this isn't super useful because we pack everything into opaque `fx.Options` groups which prevents customization, but as mentioned in the commit msg, if we stop using those and flatten everything out into a single slice of `fx.Option`, it gets way easier to customize, because `fx.Option` has a solid string representation. For example, here's the output of the option strings for the current opts:

`
fx.Options(fx.Options(fx.Provide(github.com/ipfs/go-ipfs/core/node.(*BuildCfg).options.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node.(*BuildCfg).options.func3()), fx.Provide(github.com/ipfs/go-ipfs/core/node.(*BuildCfg).options.func4()), fx.Provide(github.com/ipfs/go-ipfs/core/node.(*BuildCfg).options.func2())), fx.Provide(github.com/ipfs/go-ipfs/core/node.baseProcess()), fx.Options(fx.Provide(github.com/ipfs/go-ipfs/core/node.RepoConfig()), fx.Provide(github.com/ipfs/go-ipfs/core/node.Datastore()), fx.Provide(github.com/ipfs/go-ipfs/core/node.BaseBlockstoreCtor.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node.GcBlockstoreCtor())), fx.Options(fx.Provide(github.com/ipfs/go-ipfs/core/node.Identity.func2()), fx.Provide(github.com/ipfs/go-ipfs/core/node.PrivateKey.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.Peerstore()), fx.Invoke(github.com/ipfs/go-ipfs/core/node/libp2p.PstoreAddSelfKeys())), fx.Options(fx.Provide(github.com/ipfs/go-ipfs/core/node.RecordValidator())), fx.Options(fx.Provide(github.com/ipfs/go-ipfs/core/node.OnlineExchange.func1()), fx.Options(), fx.Provide(github.com/ipfs/go-ipfs/core/node.DNSResolver()), fx.Provide(github.com/ipfs/go-ipfs/core/node.Namesys.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node.Peering()), fx.Invoke(github.com/ipfs/go-ipfs/core/node.PeerWith.func1()), fx.Invoke(github.com/ipfs/go-ipfs/core/node.IpnsRepublisher.func1()), fx.Provide(github.com/ipfs/go-ipfs/p2p.New()), fx.Options(fx.Options(fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.init.func2()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.PNet()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.ConnectionManager()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.Host()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.MultiaddrResolver()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.DiscoveryHandler()), fx.Invoke(github.com/ipfs/go-ipfs/core/node/libp2p.PNetChecker())), fx.Supply(libp2p.AddrInfoChan), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.ResourceManager.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.AddrFilters.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.AddrsFactory.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.SmuxTransport.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.RelayTransport.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.RelayService.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.Transports.func1()), fx.Invoke(github.com/ipfs/go-ipfs/core/node/libp2p.StartListening.func1()), fx.Invoke(github.com/ipfs/go-ipfs/core/node/libp2p.SetupDiscovery.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.ForceReachability.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.HolePunching.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.Security.func2()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.Routing()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.BaseRouting.func1()), fx.Options(), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.BandwidthCounter()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.init.func3()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.AutoRelay.func1()), fx.Invoke(github.com/ipfs/go-ipfs/core/node/libp2p.AutoRelayFeeder.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.AutoNATService.func1()), fx.Provide(github.com/ipfs/go-ipfs/core/node/libp2p.ConnectionManager.func1()), fx.Options(), fx.Options()), fx.Options(fx.Options(fx.Provide(github.com/ipfs/go-ipfs/core/node.ProviderQueue()), fx.Provide(github.com/ipfs/go-ipfs/core/node.SimpleProvider()), fx.Provide(github.com/ipfs/go-ipfs-provider/simple.NewBlockstoreProvider()), fx.Provide(github.com/ipfs/go-ipfs/core/node.SimpleReprovider.func1())), fx.Options(), fx.Provide(github.com/ipfs/go-ipfs/core/node.BatchedProviderSys.func1()))), fx.Options(fx.Provide(github.com/ipfs/go-ipfs/core/node.BlockService()), fx.Provide(github.com/ipfs/go-ipfs/core/node.Dag()), fx.Provide(github.com/ipfs/go-ipfs/core/node.FetcherConfig()), fx.Provide(github.com/ipfs/go-ipfs/core/node.Pinning()), fx.Provide(github.com/ipfs/go-ipfs/core/node.Files())))
fx.WithLogger(go.uber.org/fx.glob..func1())
`

If this were flattened out instead of using `fx.Options` to group things, then we could easily e.g. remove certain `fx.Invoke()` entries. This gets even better if we only use named functions for `fx.Invoke()` instead of anonymous functions, so they look like `fx.Invoke(github.com/ipfs/go-ipfs/core/SomeExportedFunc)` instead of `fx.Invoke(github.com/ipfs/go-ipfs/core/func1()`, and they can be easily wrapped at that point. AFAIU, passing `fx.Options` is functionally equivalent to passing `[]fx.Option`.

In most cases, where you want to customize the implementation of an interface like `exchange.Interface` (for customizing Bitswap), you just append an `fx.Replace()` or `fx.Decorate()` to the opts. Plugin authors should be prefer this, as it is much less likely to break over time.

**Comment by BigLep** [comment]
2022-06-30 conversation:
* Have a way to exercise the code like install a plugin that does FX.invoke that prints something out and then check in sharness that the debug statement prints out.
* We're not using go-test because we need Make involved.
* See if we can avoid Make by using the preload function.  
* After having a way to test/verify, lets switch to "review" and "out of draft"

**Comment by BigLep** [comment]
@guseggert : I'm moving to the next iteration.  Let me know if you disagree.

File: core/builder.go
Comment (aschmahmann) [inline core/builder.go:27]: I get that we're modifying globals to work with the plugin system as we do in some other places. However, I wonder if they shouldn't get passed into `NewNode` (in the build config or as optional arguments)?

It's likely not that big a deal, but I'm mostly worried about if people use kubo as a library rather than using the constituent libraries and start messing with this function. We could either put big warning flags on it, or we could move the singleton pieces closer to the where the executable creates the `NewNode`.

File: plugin/loader/preload.go
Comment (aschmahmann) [inline plugin/loader/preload.go:24]: It's a bit awkward that we have useless test code compiled into the binary hidden with code hidden behind an env var, although not the end of the world if it's just way easier than the alternative.

**Review by aschmahmann — COMMENTED** [review]
Seems reasonable to me, left a couple comments.

We may have to evolve this or the documentation as people play around with it and see how it works, but at least there's a hook now.

File: core/builder.go
Comment (guseggert) [reply core/builder.go:27]: Do you have any examples of bad things people would do with this when using kubo as a library?

Another option is to unexport the register func, and pass it as a func into the PluginLoader.

Overall I'm biasing towards keeping it consistent w/ the rest of the plugins so the mechanics and caveats for consumers are the same.

File: plugin/loader/preload.go
Comment (guseggert) [reply plugin/loader/preload.go:24]: I completely agree but I can't think of a feasible alternative that won't be a huge time suck. This feels yucky but AFAIU in practice it has insignificant impact.

File: core/builder.go
Comment (aschmahmann) [reply core/builder.go:27]: > Do you have any examples of bad things people would do with this when using kubo as a library?

Basically if they try and instantiate two kubo nodes in the same process and get confused by global variables. The kubo as a library examples do this, they obviously don't use any custom fx options but if someone tried to fork the example and pass in a custom option things could get confusing.

If you leave it like this I'd add some comments to the function explaining the situation around multiple instantiations, even if it is ultimately just what happens when you use globals for things. Same as the problems with https://github.com/ipfs/go-unixfs/blob/4571136c15d0678006f09ebde615d0c45c126ff9/io/directory.go#L27.

kubo calls `NewNode` a few places within the codebase. Have you taken a look at them to see if it's a good idea for the custom plugins to be shared across all of those places?

> Overall I'm biasing towards keeping it consistent w/ the rest of the plugins so the mechanics and caveats for consumers are the same.

The main reason I'm flagging here to is that this plugin is a little different from the more common usages of the Datastore and IPLD Codec plugins which just add functionality the user may take advantage of. This one can change existing functionality as well.

Comment (guseggert) [reply core/builder.go:27]: Thanks for the extra info.

Yes I've taken a look at the callers of `NewNode`. There's no real way to know whether it makes sense to apply these custom fx opts to all the various places that `NewNode` is called, it depends on what the options are doing. So I think we should just apply them everywhere and tell plugin authors "your options will be applied every time a node is created", and so it's up to the plugin author to make sure that works in all contexts. There currently isn't a way for a plugin to know what "context" it's in (e.g. "am I in offline mode?")...we could pass that in to the plugin as well, something like:

```go
type NodeInfo {
	FXOpts []fx.Option
	BuildCfg *core.BuildConfig
}
FXOptions(info NodeInfo) ([]fx.Option, error)`)
```

Then the plugin could use that extra info to decide what fx opts to add. (Also, passing a struct here so we can add more fields in the future without breaking the interface).

Would that make sense?

Comment (aschmahmann) [reply core/builder.go:27]: Follow up from in person conversation:

> Would that make sense?

It could. We could also just document what's going on like you said and wait on feedback from the first few groups that try using this.

For example, we've seen a few recent cases of people wanting to build plugins we could point this way if/until something better comes along
1. Adding filtering to Bitswap server (who I'll serve data to)
2. Adding filtering to the blockservice for which blocks we'll download
3. Having a custom blockstore that fetches data from a network backend and only processes data with certain codecs

Some of these (e.g. adding allowlists in option 2) will break some node instantiations (e.g. for the node we spin up to do migrations) are potentially problematic.

A reasonable approach is probably to start with globally applied options here with some comments documenting the hazards and then we can add more data to the `NodeInfo` (better name pending) struct to help users disambiguate between what the nodes are used for.

Comment (guseggert) [reply core/builder.go]: Sounds good, I've added these caveates and warnings to the docs, and also switch the input to be a struct so that we can add more info later if we need to.

File: go.mod
Comment (aschmahmann) [inline go.mod:116]: this bump doesn't seem necessary, but also seems highly unlikely to be a problem 😄

**Review by aschmahmann — APPROVED** [review]
LGTM, let's get started here and we'll see how people start using it and evolve as needed.

Thanks for pushing this over the finish line 🙏

**Comment by aschmahmann** [comment]
@guseggert I realized after merging that while this is fine we should add some docs about this to https://github.com/ipfs/kubo/blob/master/docs/plugins.md#plugin-types. Do you mind taking care of that in a follow up PR?

**Pull request merged**
